vault/logical_system_rotate.go

```diff
@@ -0,0 +1,908 @@
+		}
+
+		secretThreshold, ok, err := data.GetOkErr("secret_threshold")
+		if err != nil {
+			return handleError(err)
+		}
+		if ok {
+			rotateConf.SecretThreshold = secretThreshold.(int)
+		}
+
+		pgpKeys, ok, err := data.GetOkErr("pgp_keys")
+		if err != nil {
+			return handleError(err)
+		}
+		if ok {
+			rotateConf.PGPKeys = pgpKeys.([]string)
+		}
+
+		backup, ok, err := data.GetOkErr("backup")
+		if err != nil {
+			return handleError(err)
+		}
+		if ok {
+			rotateConf.Backup = backup.(bool)
+		}
+
+		verificationReq, ok, err := data.GetOkErr("require_verification")
+		if err != nil {
+			return handleError(err)
+		}
+		if ok {
+			rotateConf.VerificationRequired = verificationReq.(bool)
+		}
+
+		recovery := strings.Contains(req.Path, "recovery")
+		result, err := b.Core.InitRotation(ctx, rotateConf, recovery)
+		if err != nil {
+			return handleError(err)
+		}
+
+		// this can only happen in case of autoseal recovery rotation
+		// due to keys not existing before, as the instance was initialized
+		// with recovery config secret shares set to 0
+		if result != nil {
+			keys := make([]string, 0, len(result.SecretShares))
+			keysB64 := make([]string, 0, len(result.SecretShares))
+			for _, k := range result.SecretShares {
+				keys = append(keys, hex.EncodeToString(k))
+				keysB64 = append(keysB64, base64.StdEncoding.EncodeToString(k))
+			}
+			return &logical.Response{
+				Data: map[string]interface{}{
+					"complete":              true,
+					"backup":                result.Backup,
+					"pgp_fingerprints":      result.PGPFingerprints,
+					"verification_required": result.VerificationRequired,
+					"verification_nonce":    result.VerificationNonce,
+					"keys":                  keys,
+					"keys_base64":           keysB64,
+				},
+			}, nil
+		}
+
+		return b.handleRotateInitGet()(ctx, req, data)
+	}
+}
+
+// handleRotateInitDelete handles the DELETE `/sys/rotate/root/init` and `/sys/rotate/recovery/init`
+// endpoints cancelling any in-progress rotation operations.
+func (b *SystemBackend) handleRotateInitDelete() framework.OperationFunc {
+	return func(_ context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+		recovery := strings.Contains(req.Path, "recovery")
+		if err := b.Core.CancelRotation(recovery); err != nil {
+			return handleError(err)
+		}
+		return nil, nil
+	}
+}
+
+// handleRotateUpdate handles the POST `/sys/rotate/root/update` and `/sys/rotate/recovery/update`
+// endpoints used for providing a single root key share progressing the rotation of the key.
+func (b *SystemBackend) handleRotateUpdate() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		ikey, ok, err := data.GetOkErr("key")
+		if err != nil {
+			return handleError(err)
+		}
+		if !ok || ikey.(string) == "" {
+			return handleError(errors.New("'key' must be specified in request body as JSON"))
```

Comment:
nit:

```suggestion
			return handleError(errors.New("missing required field 'key'"))
```

&c.

I suppose they are JSON request bodies, but this is more inline with other parameters. My 2c.

Reply:
I also think `GetOk(...)` is _probably_ fine, but this is OK too.